src/main/java/org/aksw/iguana/cc/worker/impl/SPARQLProtocolWorker.java

```diff
@@ -0,0 +1,331 @@
+        HttpExecutionResult result = executeHttpRequest(timeout);
+        Optional<Integer> statuscode = Optional.empty();
+        if (result.response().isPresent())
+            statuscode = Optional.of(result.response().get().statusCode());
+
+        if (result.successful()) { // 2xx
+            // process result
+            if (!responseBodyProcessor.add(result.actualContentLength().getAsLong(), result.hash().getAsLong(), result.outputStream().get())) {
+                this.responseBodybbaos = result.outputStream().get();
+            } else {
+                this.responseBodybbaos = new BigByteArrayOutputStream();
+            }
+        }
+
+        if (!result.completed() && discardOnFailure) {
+            return null;
+        }
+
+        return new ExecutionStats(
+                result.queryID(),
+                result.requestStart(),
+                result.duration(),
+                statuscode,
+                result.actualContentLength(),
+                result.hash,
+                result.exception()
+        );
+    }
+
+
+    private HttpExecutionResult executeHttpRequest(Duration timeout) throws IOException, URISyntaxException {
+        final QueryHandler.QueryStreamWrapper queryHandle = config().queries().getNextQueryStream();
+        final HttpRequest request = requestFactory.buildHttpRequest(
+                queryHandle.queryInputStream(),
+                timeout,
+                config().connection(),
+                config().acceptHeader()
+        );
+
+        final Instant requestStart = Instant.now();
+        BiFunction<HttpResponse<InputStream>, Exception, HttpExecutionResult> createFailedResult = (response, e) -> {
+            final Duration requestDuration = Duration.between(requestStart, Instant.now());
+            return new HttpExecutionResult(
+                    queryHandle.index(),
+                    Optional.ofNullable(response),
+                    requestStart,
+                    requestDuration,
+                    Optional.empty(),
+                    OptionalLong.empty(),
+                    OptionalLong.empty(),
+                    Optional.ofNullable(e)
+            );
+        };
+
+        try {
+            return httpClient.sendAsync(request, HttpResponse.BodyHandlers.ofInputStream())
+                    .thenApply(httpResponse -> {
+                        try (final var bodyStream = httpResponse.body()) {
+                            if (httpResponse.statusCode() / 100 == 2) { // Request was successful
+                                OptionalLong contentLength = httpResponse.headers().firstValueAsLong("Content-Length");
+                                try (var hasher = hasherFactory.newStreamingHash64(0)) {
+                                    int readBytes;
+                                    while ((readBytes = bodyStream.readNBytes(this.buffer, 0, this.buffer.length)) != 0) {
+                                        hasher.update(this.buffer, 0, readBytes);
```

Comment:
Check in the look for timeouts.

Reply:
I'm assuming you're suggesting that, because timeouts after the response has been received should be ignored, right?

Reply:
First of all: in thenApply we have the httpResponse object but the body was not yet fully received. This is a special property of the httpClient where the result is returned as InputStream BodyHandler. Consequently, we must keep our own clock for the timeout running until we have fully consumed or the timeout is reached. That's also why we should check for timeout in the while loop.

Why should we keep things it in thenApply: readNBytes is blocking. So, it waits until either the requested amount of data is available or the stream is closed. So, if a client stalls but doesn't cancel the HTTP response, readNBytes will block. This also means that the thread will go on running even if the timeout set in `.get(timeout.toNanos(), TimeUnit.NANOSECONDS);` is reached because async threads cannot be terminated unless they finish their work or the pool is shut down. SO, with the `get` method, the calling thread is waken up but the thread that was is still running. That is also why we need to put a check for an timeout into the loop reading the stream: if reached return failure or throw. 

And, that's also the reason why we must check with `pool.getActiveCount() < 2` if enough threads are available before we send the next request. Otherwise, we will wait for those threads to terminate before our workload starts.

Reply:
I gave the httpclient in the sparqlworker a seperate ExecutorService. With that it's possible to forcefully shutdown a running task before the next request is sent.